scheduler/src/cook/kubernetes/controller.clj

```diff
@@ -0,0 +1,219 @@
+    {:expected-state :expected/completed}))
+
+(defn process
+  "Visit this pod-name, processing the new level-state. Returns the new expected state. Returns
+  empty dictionary to indicate that the result should be deleted. NOTE: Must be invoked with the lock."
+  [{:keys [api-client existing-state-map expected-state-map] :as kcc} pod-name]
+  (loop [{:keys [expected-state] :as expected-state-dict} (get @expected-state-map pod-name)
+         {:keys [synthesized-state pod] :as existing-state-dict} (get @existing-state-map pod-name)]
+    ;; TODO: Remove the printing of existing-state-dict once we test on real kubernetes and get synthesized-pod-state robust.
+    (log/info "Processing: " pod-name ": ((" (prepare-expected-state-dict-for-logging expected-state-dict)
+              " ===== " (or (:state synthesized-state) :missing)  "///" existing-state-dict  "))")
+    ; TODO: We added an :expected/starting state to the machine, to represent when a pod is starting. We map instance.status/unknown to that state
+    ; The todo is to add in cases for [:expected/starting *] for those other states.
+    (let
+      [new-expected-state-dict (case (vector (or expected-state :missing) (or (:state synthesized-state) :missing))
+                                 [:expected/starting :missing] (launch-task api-client expected-state-dict)
+                                 [:expected/starting :pod/running] (pod-has-started kcc existing-state-dict)
+                                 ; TODO We need to add cases for the other [:expected/starting *] states in.
+                                 [:expected/starting :pod/waiting] expected-state-dict
+                                 [:expected/starting :pod/succeeded] (pod-has-just-completed kcc existing-state-dict)
+                                 [:expected/starting :pod/failed] (pod-has-just-completed kcc existing-state-dict)
+                                 [:expected/running :pod/running] expected-state-dict
+                                 [:expected/running :pod/succeeded] (pod-has-just-completed kcc existing-state-dict)
+                                 [:expected/completed :pod/succeeded] (remove-finalization-if-set-and-delete api-client expected-state-dict pod)
+                                 [:expected/running :pod/failed] (pod-has-just-completed kcc existing-state-dict)
+                                 [:expected/completed :pod/failed] (remove-finalization-if-set-and-delete api-client expected-state-dict pod)
+                                 [:expected/completed :missing] nil ; Cause it to be deleted.
+                                 [:expected/killed :pod/waiting] (kill-task api-client expected-state-dict pod)
+                                 [:expected/killed :pod/running] (kill-task api-client expected-state-dict pod) ; TODO: Where does the datomic update occur? Do we do it when we do [expected/killed :pod/failed], to be similar to mesos, we update it only when the backend says its dead?
```

Comment:
Is this TODO now dead with the handle-status-update refactor?

Reply:
👍